packages/chain-adapters/src/sui/SuiChainAdapter.ts

```diff
@@ -512,7 +515,102 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SuiMainnet> {
     return
   }
 
-  parseTx(): Promise<never> {
-    throw new Error('SUI transaction parsing not yet implemented')
+  async parseTx(txHashOrTx: unknown, pubkey: string): Promise<Transaction> {
+    try {
+      // Fetch full transaction data if only txHash was provided
+      const tx =
+        typeof txHashOrTx === 'string'
+          ? await this.client.getTransactionBlock({
+              digest: txHashOrTx,
+              options: {
+                showInput: true,
+                showEffects: true,
+                showBalanceChanges: true,
+              },
+            })
+          : (txHashOrTx as SuiTransactionBlockResponse)
+
+      const sender = tx.transaction?.data.sender ?? ''
+
+      const txid = tx.digest
+      const blockHeight = Number(tx.checkpoint ?? 0)
+      const blockTime = tx.timestampMs ? Math.floor(Number(tx.timestampMs) / 1000) : 0
+
+      const latestCheckpoint = await this.client.getLatestCheckpointSequenceNumber()
+      const confirmations = tx.checkpoint ? Number(latestCheckpoint) - Number(tx.checkpoint) + 1 : 0
+
+      const status =
+        tx.effects?.status.status === 'success'
+          ? TxStatus.Confirmed
+          : tx.effects?.status.status === 'failure'
+          ? TxStatus.Failed
+          : TxStatus.Unknown
+
+      const gasUsed = tx.effects?.gasUsed
+      const fee = gasUsed
+        ? {
+            assetId: this.assetId,
+            value: (
+              BigInt(gasUsed.computationCost) +
+              BigInt(gasUsed.storageCost) -
+              BigInt(gasUsed.storageRebate)
+            ).toString(),
+          }
+        : undefined
+
+      const balanceChanges = tx.balanceChanges ?? []
+
+      const transfers = balanceChanges.map(change => {
+        let ownerAddress: string | null = null
+        if (typeof change.owner === 'object' && 'AddressOwner' in change.owner) {
+          ownerAddress = change.owner.AddressOwner
+        }
+
+        const assetId =
+          change.coinType === '0x2::sui::SUI'
+            ? this.assetId
+            : toAssetId({
+                chainId: this.chainId,
+                assetNamespace: ASSET_NAMESPACE.suiCoin,
+                assetReference: change.coinType,
+              })
+
+        const amount = BigInt(change.amount)
+        const isReceive = amount > 0n
+        const isSend = amount < 0n
+
+        const transferType =
+          ownerAddress === pubkey
+            ? isReceive
+              ? TransferType.Receive
+              : TransferType.Send
+            : TransferType.Contract
+
+        return {
+          assetId,
+          from: isSend ? [sender] : ownerAddress ? [ownerAddress] : [sender],
+          to: isReceive ? [ownerAddress ?? sender] : [sender],
+          type: transferType,
+          value: amount < 0n ? (-amount).toString() : amount.toString(),
```

Comment:
Probably not the clean code area those days but we can always use bnOrZero if we want